### ISSUE TITLE
Add support for SQS redrives

### ISSUE THREAD

**Issue by rpk** [spring-cloud/spring-cloud-aws#30]
Currently, we can't find a way to prevent the message from being deleted after it was received.  Generally, when a message fails to process (an exception is thrown from the message handler) the message should be left in the queue to be redriven.  [SQS Redrive Policy with DLQ](http://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/SQSDeadLetterQueue.html)

In executeMessage catching throwable prevents the interruption of deleteMessage:([SimpleMessageListenerContainer.java#L125-L131](https://github.com/spring-cloud/spring-cloud-aws/blob/master/spring-cloud-aws-messaging/src/main/java/org/springframework/cloud/aws/messaging/listener/SimpleMessageListenerContainer.java#L125-L131))
```java
protected void executeMessage(org.springframework.messaging.Message<String> stringMessage) {
	try {
		getMessageHandler().handleMessage(stringMessage);
	} catch (Throwable throwable) {
		handleError(throwable);
	}
}
```

([SimpleMessageListenerContainer.java#L191-L203](https://github.com/spring-cloud/spring-cloud-aws/blob/master/spring-cloud-aws-messaging/src/main/java/org/springframework/cloud/aws/messaging/listener/SimpleMessageListenerContainer.java#L191-L203))

```java
		@Override
		public void run() {
			String receiptHandle = this.message.getReceiptHandle();
			String payload = this.message.getBody();
			MessageBuilder<String> messageBuilder = MessageBuilder.
					withPayload(payload).
					setHeader(QueueMessageHandler.Headers.LOGICAL_RESOURCE_ID_MESSAGE_HEADER_KEY, this.logicalQueueName);
			copyAttributesToHeaders(messageBuilder);
			executeMessage(messageBuilder.build());
			getAmazonSqs().deleteMessage(new DeleteMessageRequest(this.queueUrl, receiptHandle));
			getLogger().debug("Deleted message with id {} and receipt handle {}", this.message.getMessageId(), this.message.getReceiptHandle());
		}
	}
```

Ideally letting an uncaught exception be thrown from the task should prevent the message from being deleted (it will become available again after the visibility timeout has elapsed).  Alternatively, throwing an exception inherited from a newly introduced exception (e.g. RedrivableException) that was handled before the throwable would work as well.

**Comment by alainsahli** [comment]
Hi @rpk 

Thanks for you input!

We are going to modify the handling of failing messages. If the queue is marked as "Use Redrive Policy" then the failing messages won't be deleted. Therefore it will be possible to create a new message mapping on the dead letter queue and handle the failing messages accordingly.

Would this approach solve your problem?

**Comment by rpk** [comment]
Yep, that would work for us.

**Comment by alainsahli** [comment]
Resolved with 07f02fec1dfee3dc39ea117eff1a5a85a58147bd
